cqf-fhir-utility/src/main/java/org/opencds/cqf/fhir/utility/visitor/dstu3/KnowledgeArtifactPackageVisitor.java

```diff
@@ -175,4 +209,396 @@ public static List<BundleEntryComponent> findUnsupportedInclude(
+            ValueSet.ValueSetExpansionParameterComponent parameterNaive =
+                    new ValueSet.ValueSetExpansionParameterComponent();
+            parameterNaive.setName("naive");
+            parameterNaive.setValue(new BooleanType(true));
+            expansionParams.add(parameterNaive);
+            expansion.setParameter(expansionParams);
+
+            for (ValueSet.ConceptSetComponent csc : valueSet.getCompose().getInclude()) {
+                for (ValueSet.ConceptReferenceComponent crc : csc.getConcept()) {
+                    expansion
+                            .addContains()
+                            .setCode(crc.getCode())
+                            .setSystem(csc.getSystem())
+                            .setVersion(csc.getVersion())
+                            .setDisplay(crc.getDisplay());
+                }
+            }
+            valueSet.setExpansion(expansion);
+        } else {
+            Optional<StringType> username = Optional.empty();
+            Optional<StringType> apiKey = Optional.empty();
+            if (terminologyEndpoint.isPresent()) {
+                Endpoint endPnt = (Endpoint) terminologyEndpoint.get();
+                username = endPnt.getHeader().stream()
+                        .filter(h -> h.getId().equals("username"))
+                        .findFirst();
+                apiKey = endPnt.getHeader().stream()
+                        .filter(h -> h.getId().equals("apiKey"))
+                        .findFirst();
+            }
+
+            if (!username.isPresent() || !apiKey.isPresent()) {
+                throw new UnprocessableEntityException(
+                        "Cannot expand ValueSet without credentials: " + valueSet.getId());
+            }
+            try {
+                expandedValueSet = terminologyServerClient.expand(
+                        valueSet,
+                        authoritativeSourceUrl,
+                        expansionParameters,
+                        username.get().getValue(),
+                        apiKey.get().getValue());
+                valueSet.setExpansion(expandedValueSet.getExpansion());
+            } catch (Exception ex) {
+                System.out.println("Terminology Server expansion failed: {"
+                        + valueSet.getIdElement().getValue() + "}");
+            }
+        }
+    }
+
+    public boolean isVSMAuthoredValueSet(ValueSet valueSet) {
+        return valueSet.hasMeta()
+                && valueSet.getMeta().hasTag()
+                && valueSet.getMeta()
+                                .getTag(
+                                        KnowledgeArtifactAdapter.vsmValueSetTagCodeSystemUrl,
+                                        KnowledgeArtifactAdapter.vsmValueSetTagVSMAuthoredCode)
+                        != null;
+    }
+
+    public boolean hasSimpleCompose(ValueSet valueSet) {
+        if (valueSet.hasCompose()) {
```

Comment:
Can't all this be simplified as something like:

return valueSet.asCompose() && 
           !valueSet.getCompose().hasExclude && 
           valueSet.getCompose().getInclude().noneMatch(cdc -> cdc.hasValueSet() || !cdc.hasSystem() || ...)

thinking on top of my head so there may be some && that are ||, etc...

Reply:
Resolved! Thanks